File: boj24337_가희와탑.java
import java.util.*;
import java.io.*;
import java.math.*;

//time=1sec
//memory=512mb
//n= 1 <= n <= 100_000

public class boj24337_가희와탑 {
    static int n, a, b;
    static ArrayDeque<Integer> building;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        String[] input = br.readLine().split(" ");
        n = Integer.parseInt(input[0]);
        a = Integer.parseInt(input[1]);
        b = Integer.parseInt(input[2]);
        building = new ArrayDeque<Integer>();
        int cnt = 0;
        int curr = 0;

        if (a+b > n+1) {
            System.out.println(-1);
            return ;
        }

        if (a>b) {
            for (int i = 1; i <= a; ++i) {
                building.addLast(i);
            }
            if (b==1) {
                cnt = n-a;
                while (cnt > 0) {
                    building.addFirst(1);
                    --cnt;
                }
            } else {
                curr = b-1;
                cnt = b-1;
                while (building.size() < n && curr > 0) {
                    building.addLast(curr);
                    --curr;
                    --cnt;
                }
                while (building.size() < n) {
                    building.addFirst(1);
                }
            }
            // System.out.println(building);
        } else {

            if (a == 1) {
                cnt = n-b;
                for (int i = b-1; i >= 1; --i) {
                    building.addLast(i);
                }
                while (cnt > 0) {
                    building.addFirst(1);
                    --cnt;
                }
                building.addFirst(b);
            } else {
                for (int i = b; i>= 1; --i) {
                    building.addLast(i);
                }
                cnt = a-1;
                curr = a-1;
                while (building.size() < n && curr > 0) {
                    building.addFirst(curr);
                    --cnt;
                    --curr;
                }
                while (building.size() < n) {
                    building.addFirst(1);
                }
            }
            // System.out.println(building);
        }
        StringBuilder ans = new StringBuilder();
        while (!building.isEmpty()) {
            ans.append(building.peek());
            ans.append(" ");
            building.pop();
        }
        bw.write(ans.toString());
        bw.flush();
        bw.close();
    }
}

File: boj24337_가희와탑.kt
import kotlin.math.*

//time=1sec
//memory=512mb
//n= 1 <= n <= 100_000


class Boj24337 {
    val n: Int
    val a: Int
    val b: Int
    val buildings: ArrayDeque<Int>
    val ans: MutableList<Int>

    init {
        readln().split(' ').map { it.toInt() }.also {
            n = it[0]
            a = it[1]
            b = it[2]
        }
        buildings = ArrayDeque()
        ans = mutableListOf()
    }

    fun solution() {
        // var curr = 0
        // var cnt = n

        if (a + b > n + 1) {
            println(-1)
            return
        }
        var highest = max(a, b)
        if (a>b) {
            for (i in 1..highest) {
                buildings += i
            }
            if (b == 1) {
                repeat(n-a) {
                    buildings.addFirst(1)
                }
            } else {
                var curr = b-1
                while (buildings.size < n && curr > 0) {
                    buildings += curr
                    --curr
                }
                while (buildings.size < n) {
                    buildings.addFirst(1)
                }
            }
            buildings.forEach {
                print("$it ")
            }
        } else {
            for (i in highest downTo 1) {
                buildings += i
            }
            if (a == 1) {
                repeat(n-b) {
                    buildings.add(1, 1)
                }
            } else {
                var curr = a-1
                while (buildings.size < n && curr > 0) {
                    buildings.addFirst(curr)
                    --curr
                }
                while (buildings.size < n) {
                    buildings.addFirst(1)
                }
            }
            buildings.forEach {
                print("$it ")
            }
        }
    }
}

fun main() {
    Boj24337().solution()
}